docs/specs/future-of-workload.md

```diff
@@ -0,0 +1,43 @@
+
+Also, there is no real need of including the Aspire.Hosting package as part of the workload, as there is no guarantee that this is the version that the AppHost project will want to use, therefore we would also remove this package from the workload. In the end, the workload will only contain the Aspire.Hosting.SDK and Aspire.ProjectTemplates components, which are the ones that can be truly considered as tooling components, and in both cases you would want to always be in the latest version. These two are also not tightly coupled to the version of the Aspire.Hosting package that the AppHost project references, so they can be updated independently of the AppHost.
+
+This is the proposed new design:
+
+![Proposal](workload-proposed-state.png "Proposal")
+
+### Pros
+
+This proposal has many benefits which will solve the issues we currently have with the workload:
+
+- The AppHost project will have full control over the version of the DCP and Dashboard that it wants to use.
+- Customers can have multiple AppHost projects that reference different versions of Aspire.Hosting without any issues.
+- We can update the DCP and Dashboard independently of the workload, which will allow us to ship updates to these components more frequently and with less risk.
+- Customers can stay in the latest version of the workload without worrying about compatibility issues with the AppHost project.
+
+### Cons
+
+- The main drawback of removing DCP and Dashboard from the workload, is the potential impacts it may have with restore times and local package caches. DCP especially is not small in size (~150mb) and it is also a native executable (written in Go), so we currently have separate packages each carrying the executable for each individual platform. One thing we were getting for free with DCP being part of the workload, is that the SDK authoring would automatically only install the DCP package for the platform that the SDK is being installed on. NuGet doesn't have great support today for platform-specific dependencies, which means that with the above proposal the Aspire.Hosting.AppHost package would depend on all of the DCP packages for all platforms, which would increase the size of the things that need to be restored, as well as the time it takes to do it.
```

Comment:
Could we not utilize the ref-split packages pattern to achieve the same outcome of only downloading the runtime packages required for the current runtime? IIUC this is exactly how the .NET SDK itself avoids downloading runtime packages for shared frameworks until specifically required.

Reply:
Sorry, I'm not sure I'm following here. Shared frameworks are kind of special here as they are built-in on the SDK itself, and one of the "implicit" references, so the SDK has special logic to deal with that and not have to pull in all different platform specific runtime packs. For regular NuGet packages, AFAIK there isn't a NuGet built-in way of restoring platform-specific packages unless you restore using a RID, which IMO is not something we want to enforce on our customers. That's how the whole runtime.json stuff for RID specific dependencies worked, it basically relied on people doing `dotnet restore -r  win-x64` so the right section of the runtime.json could be picked.

Reply:
I just meant using the pattern of splitting different platform implementations into `runtime.[rid].` prefixed packages and then having our targets in `Aspire.Hosting` take care of restoring them and injecting the appropriate paths into the AppHost assembly. I think this is very similar to how things like the ILCompiler toolchain works today, e.g. [runtime.win-x64.Microsoft.DotNet.ILCompiler](https://www.nuget.org/packages/runtime.win-x64.Microsoft.DotNet.ILCompiler/)

Reply:
(Note: I'm not an expert here, so I could be wrong)

The ILCompiler and crossgen are "special" because they are known by the SDK explicitly.

https://github.com/dotnet/sdk/blob/e80b6ebc8b92ce2026f82cbfa34257d05d58cf83/src/Installer/redist-installer/targets/GenerateBundledVersions.targets#L553-L565

And then that `**RID**` gets replaced with the correct RID in a C# task:

https://github.com/dotnet/sdk/blob/e80b6ebc8b92ce2026f82cbfa34257d05d58cf83/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L789

I'm not sure we could take this same approach outside of the SDK.

There is/was a feature in NuGet that we used to use called `runtime.json`, but AFAIK it was never officially supported outside of System.* packages (which stopped using the feature).

See:

* https://github.com/NuGet/Home/issues/6083
* https://github.com/NuGet/Home/issues/10571
* https://github.com/NuGet/Home/issues/1660
* https://github.com/NuGet/Home/issues/3114

Reply:
I also believe this is not something that any package can do, but I'll look into it.

Reply:
I believe you're right @eerhardt, but why can't we take a similar approach? What is it that the SDK does from its MSBuild targets that's so special that we can't do it from targets in `Aspire.Hosting.AppHost`? We only need to facilitate the download of the required runtime package at the right time. Things like `dotnet new install` do this today, as well as the SDK itself of course for scenarios like self-contained publish. Once the runtime-specific package is downloaded its path can be injected into the AppHost app via its targets.

Reply:
I may be missing something here, but the main difference would be that this SDK logic (because it is in the SDK) is able to pick the right RID and add the right PackageReference _before_ restore happens. If we were to add something like this in our Aspire.Hosting.AppHost targets, by definition those cannot run at restore time (as they need to be restored in the initial phase), meaning the only way to add the PackageReferences for the specific RID packages from there would be at Build time (as opposed to Restore time) which is a big no-no. Restore graph should never be modified post initial-restore as a lot of things break if at build time you are trying to add new references.

Reply:
Sorry, I'm not saying you do the restore as PackageReferences, that's not required if you're just trying to ensure packages are downloaded. These are tools, there's nothing in them that needs to be referenced by the app. We just need them on the machine, so manually restoring at some other point is fine.

Reply:
I see. How would we find the right path to the tools then? The other thing that these packages have is a targets file that sets the path to the tools relative to the targets file itself, and the targets are only loaded because this is a PackageReference. If we don't load these targets and instead assume what the right path to the restored tools are, we may run into issues when customers don't use default restore path, which I suppose is not terribly un-common.

Reply:
> What is it that the SDK does from its MSBuild targets that's so special that we can't do it from targets in Aspire.Hosting.AppHost?

It gets to inject logic that influences restore. That Task I referenced above gets run before restore happens.

When would the download of DCP/Dashboard actually happen?

1. During `dotnet restore` - this is when the `Aspire.Hosting.AppHost` nupkg would get restored. Note that MSBuild logic inside that nupkg can't run during restore because it isn't there yet - it hasn't been restored.
2. During `dotnet build` - here we CAN run logic from inside the `Aspire.Hosting.AppHost` nupkg. So we can download the nupkgs for DCP/Dashboard. But note that the MSBuild logic from inside those nupkgs can't influence the build anymore, because the build is already happening. I'm not sure there is a way to dynamically load more MSBuild logic once the build as started.

So to make it work this way with `Aspire.Hosting.AppHost`, we would need to download the executable packages during build - and we would need to have the logic about where the executables exist inside those nupkgs in `Aspire.Hosting.AppHost`.

To actually do the download during restore, we would need to inject MSBuild logic into the project that influences restore - i.e. either in the workload as we do today, or in an MSBuild Sdk.

Reply:
I'll chat with @rainersigwald about this to get his feedback. Even if they are not package references and  we are just downloading the tools, hitting the network at build time is typically an anti-pattern. I'm also concerned with the assumptions we'd have to make from `Aspire.Hosting.AppHost` package targets to be able to find the restored path to the tools.

Reply:
We'd only perform the download if required and the assumptions about paths could be avoided by controlling where the download is made to (again, just like `dotnet new install` does).

Reply:
I don't think doing the download during build is going to be a great user experience. I also don't think we want to sign up for handling the download of these packages ourselves. That seems like a bug farm.

If we are going with RID-split packages, the best option (IMO) is to get NuGet to manage the downloading/restoring/caching/etc, which means getting our logic injected so it can influence restore.

Reply:
> If we are going with RID-split packages, the best option (IMO) is to get NuGet to manage the downloading/restoring/caching/etc, which means getting our logic injected so it can influence restore.

Yeah I can't argue with that. Is it as simple as having our SDK targets set `<UseCurrentRuntimeIdentifier>` to "true" before restore, and having `Aspire.Hosting.AppHost` depend on `Aspire.Hosting.Orchestration`, which in turn contains a `runtime.json` file that defines the runtime-specific packages to bring in?

Reply:
There is https://learn.microsoft.com/en-us/nuget/consume-packages/packagedownload-functionality also. This will just download the specific nuget as part of the regular restore.

`dotnet/runtime` uses this for wasm in a couple of places.

Reply:
>  Is it as simple as having our SDK targets set <UseCurrentRuntimeIdentifier> to "true" before restore, and having Aspire.Hosting.AppHost depend on Aspire.Hosting.Orchestration, which in turn contains a runtime.json file that defines the runtime-specific packages to bring in?

I don't think we want to use `runtime.json`. See https://github.com/dotnet/runtime/issues/49137 where the runtime is trying to get off that feature.

Reply:
CC @dsplaisted who's our expert on packagedownload as well as the special processing we do for RIDs for framework packages.

Reply:
I don't believe the existing NuGet runtime.json / RID-specific functionality would work for this at all (even if we weren't trying to avoid using it in general).  That feature lets you get different packages based on the RuntimeIdentifier the app is using.  However, what we want here is to get the DCP etc. packages for the RuntimeIdentifier of the host.  If we support cross-targeting (ie building an Aspire app for linux-x64 from a windows-x64 machine), then NuGet's runtime.json functionality doesn't help.

It may be possible to avoid downloading DCP etc. for all platforms with logic in the Aspire SDK targets.  There could be logic that detected a PackageReference to `Aspire.Hosting.AppHost`, and if it was found would add `PackageReference` or `PackageDownload` items to the corresponding platform-specific versions of the DCP packages.  That might have some issues and wouldn't work for transitive references to `Aspire.Hosting.AppHost`, but if I understand correctly there may not be scenarios where that package is transitively referenced.

Reply:
> If we support cross-targeting (ie building an Aspire app for linux-x64 from a windows-x64 machine), then NuGet's runtime.json functionality doesn't help.

The AppHost project is never deployed so I don't think this is an issue.

> It may be possible to avoid downloading DCP etc. for all platforms with logic in the Aspire SDK targets. There could be logic that detected a PackageReference to Aspire.Hosting.AppHost, and if it was found would add PackageReference or PackageDownload items to the corresponding platform-specific versions of the DCP packages. That might have some issues and wouldn't work for transitive references to Aspire.Hosting.AppHost, but if I understand correctly there may not be scenarios where that package is transitively referenced.

This is the approach being proposed now I believe.